streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractStateManager.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.internals;
+
+import java.util.Map;
+
+public abstract class AbstractStateManager implements StateManager {
```

Comment:
I know you've already had a conversation about this in the last cycle of reviews and defended the idea. So, I'm sorry to say that I also disagree with adding this class.

We've recently spent several months in an extremely costly refactoring effort clearing up a bunch of unmaintainable code in this exact module. There were many flaws that we had to correct, but one of the key ones was certainly excessive abstraction. There's nothing excessive about this class _right now_, but experience says it will become so over time. I'd avoid adding abstract classes unless there's an extremely compelling reason to do it. Right now, it just looks like we're saving the concrete classes from storing one field, which doesn't seem too compelling.

I really want to emphasize that your choice to add this class is very normal, and in some codebases, it would be required style. But, given my experience with maintaining this codebase, I'd really like to inline this and keep the class hierarchy absolutely as flat as possible.

Reply:
Since two people now criticized this, I will remove this abstract class. However, I strongly disagree that we should not abstract (and reduce code duplications) because maybe over time the abstraction might become excessive and hence unmaintainable. It is our responsibility to avoid the latter while keeping a clean design.

Reply:
Thanks! This seems like the kind of debate that will continue for a while :) 

Just to share the thought, adding abstraction isn't the only way to de-duplicate code. In fact composition is generally less costly to maintain, for example adding utility methods.